### PR TITLE
Fix "Mismatch of dweight at layernorm_backward.cu #428"

### LINKED ISSUE #428: Mismatch of dweight at layernorm_backward.cu
with the following script

nvcc -O3 --use_fast_math -lcublas -lcublasLt layernorm_backward.cu -o layernorm_backward
./layernorm_backward 2

output
```
Using kernel 2
Checking correctness...
dinp:
-1.182338 -1.187500
0.236102 0.236328
0.667884 0.667969
-1.111703 -1.117188
0.685677 0.687500
dweight:
8.187102 4.125000
Mismatch of dweight at 0: CPU_ref: 8.187102 vs GPU: 4.125000
90.587265 104.000000
Mismatch of dweight at 1: CPU_ref: 90.587265 vs GPU: 104.000000
-38.522667 -39.750000
114.101608 108.500000
58.895130 64.500000
Mismatch of dweight at 4: CPU_ref: 58.895130 vs GPU: 64.500000
Mismatch of dweight at 7: CPU_ref: -5.614801 vs GPU: -7.562500
Mismatch of dweight at 9: CPU_ref: 67.834213 vs GPU: 91.000000
Mismatch of dweight at 10: CPU_ref: -1.165685 vs GPU: 4.500000
Mismatch of dweight at 11: CPU_ref: 131.065735 vs GPU: 98.500000
Mismatch of dweight at 17: CPU_ref: -19.924236 vs GPU: -15.812500
Mismatch of dweight at 19: CPU_ref: -31.724333 vs GPU: -28.625000
Mismatch of dweight at 21: CPU_ref: 24.697800 vs GPU: 29.000000

```

### PR DESCRIPTION
**Pull request by lancerts**
1. Fix #428, the root cause: we are doing accumulation using the bfloat precision directly in atomicaddX, which results in a final error that is large.  The fix is to do accumulation in `float` (temp array allocated on device) and write the final accumulated result as `bfloat`.
2. spell the uint to `unsigned int` since uint isn't an actual type in C or C++ (comments from @ngc92 ).

After fix, kernel 2 passed.
```
Using kernel 2
Checking correctness...
dinp:
-0.559983 -0.562500
0.620476 0.621094
0.028621 0.028564
0.765740 0.765625
0.062205 0.062256
dweight:
-23.181511 -23.125000
-85.985672 -86.000000
-16.258125 -16.250000
-3.262060 -3.156250
59.474670 59.250000
dbias:
-30.434944 -30.500000
-93.241646 -93.500000
-41.022255 -41.000000
-13.087867 -13.125000
-56.647667 -56.500000
All results match for block_size=32.

Checking correctness...
dinp:
-0.559983 -0.562500
0.620476 0.621094
0.028621 0.028564
0.765740 0.765625
0.062205 0.062256
dweight:
-23.181511 -23.125000
-85.985672 -86.000000
-16.258125 -16.250000
-3.262060 -3.156250
59.474670 59.250000
dbias:
-30.434944 -30.500000
-93.241646 -93.500000
-41.022255 -41.000000
-13.087867 -13.125000
-56.647667 -56.500000
All results match for block_size=64.

Checking correctness...
dinp:
-0.559983 -0.562500
0.620476 0.621094
0.028621 0.028564
0.765740 0.765625
0.062205 0.062256
dweight:
-23.181511 -23.125000
-85.985672 -86.000000
-16.258125 -16.250000
-3.262060 -3.156250
59.474670 59.250000
dbias:
-30.434944 -30.500000
-93.241646 -93.500000
-41.022255 -41.000000
-13.087867 -13.125000
-56.647667 -56.500000
All results match for block_size=128.

Checking correctness...
dinp:
-0.559983 -0.562500
0.620476 0.621094
0.028621 0.028564
0.765740 0.765625
0.062205 0.062256
dweight:
-23.181511 -23.125000
-85.985672 -86.000000
-16.258125 -16.250000
-3.262060 -3.156250
59.474670 59.250000
dbias:
-30.434944 -30.500000
-93.241646 -93.500000
-41.022255 -41.000000
-13.087867 -13.125000
-56.647667 -56.500000
All results match for block_size=256.

Checking correctness...
dinp:
-0.559983 -0.562500
0.620476 0.621094
0.028621 0.028564
0.765740 0.765625
0.062205 0.062256
dweight:
-23.181511 -23.125000
-85.985672 -86.000000
-16.258125 -16.250000
-3.262060 -3.156250
59.474670 59.250000
dbias:
-30.434944 -30.500000
-93.241646 -93.500000
-41.022255 -41.000000
-13.087867 -13.125000
-56.647667 -56.500000
All results match for block_size=512.

Checking correctness...
dinp:
-0.559983 -0.562500
0.620476 0.621094
0.028621 0.028564
0.765740 0.765625
0.062205 0.062256
dweight:
-23.181511 -23.125000
-85.985672 -86.000000
-16.258125 -16.250000
-3.262060 -3.156250
59.474670 59.250000
dbias:
-30.434944 -30.500000
-93.241646 -93.500000
-41.022255 -41.000000
-13.087867 -13.125000
-56.647667 -56.500000
All results match for block_size=1024.

block_size   32 time 1.3195 ms
block_size   64 time 0.7869 ms
block_size  128 time 0.5699 ms
block_size  256 time 0.4764 ms
block_size  512 time 0.4901 ms
block_size 1024 time 0.5541 ms
```

### CODE REVIEW

File: dev/cuda/layernorm_backward.cu [hunk 1]
```diff
@@ -20,7 +20,7 @@ version 2 moves a lot of reduction to shared memory over global memory
 
 #define ENABLE_BF16
 #include "common.h"
-
+//typedef unsigned int uint;       // required on windows
```
Comment (ngc92) [inline dev/cuda/layernorm_backward.cu:23]: maybe we should just spell the type correctly; uint isn't an actual type in C or C++

File: dev/cuda/layernorm_backward.cu [hunk 2]
```diff
@@ -1262,7 +1270,15 @@ void layernorm_backward2(Tdinp* dinp, Tparams* dweight, Tparams* dbias,
...
     size_t shared_mem_size = 2 * C * sizeof(float);
-    layernorm_backward_kernel2<<<grid_size, block_size, shared_mem_size>>>(dinp, dweight, dbias, dout, inp, weight, mean, rstd, B, T, C);
+    float* dweight_tmp;
+    float* dbias_tmp;
+    cudaCheck(cudaMalloc(&dweight_tmp, C * sizeof(float)));
+    cudaCheck(cudaMalloc(&dbias_tmp, C * sizeof(float)));
+    cudaMemset(dweight_tmp, 0, C * sizeof(float));
+    cudaMemset(dbias_tmp, 0, C * sizeof(float));
+    layernorm_backward_kernel2<<<grid_size, block_size, shared_mem_size>>>(dinp, dweight, dbias, dout, inp, weight, mean, rstd, B, T, C, dweight_tmp, dbias_tmp);
+    cudaDeviceSynchronize();
```
Comment (ngc92) [inline dev/cuda/layernorm_backward.cu:1280]: why is there a sync here?

Comment (lancerts) [reply dev/cuda/layernorm_backward.cu]: Shouldnt be there, removed

File: dev/cuda/layernorm_backward.cu (hunk 1 above)
Comment (lancerts) [reply dev/cuda/layernorm_backward.cu]: Good suggestion, let me do it.

**Comment by foreverpiano** [comment]
yes, i think it's precision issue. thanks for your help

**Pull request merged**